src/components/boolean-search/boolean-table.tsx

```diff
@@ -0,0 +1,307 @@
+  };
+
+  const handleAddRow = () => {
+    setRows([...rows, defaultValues]);
+  };
+
+  /**
+    * Handles changing the input field on change of rows
+  */
+  useEffect(() => {
+    if (rows.length === 0 && stageFrom.length > 0) {
+      handleAddRow();
+    }
+    setActiveRow(rows.length - 1);
+    changeFiltersDisplayText();
+  }, [rows, stageFrom]);
+
+  /**
+    * To add new row on click add button
+  */
+  useEffect(() => {
+    setActiveRow(rows.length - 1);
+    setAddClicked(false);
+  }, [addClicked]);
+
+  /**
+    * setting the left side pane source name to the input field on the right side
+  */
+  useEffect(() => {
+    // Implement the setSourceForFilter function in your React component
+    const setSourceForFilter = (sourceObject: any) => {
+      // Handle the source object and perform necessary actions
+      const updatedRows = [...rows];
+      const updatedRow = { ...updatedRows[activeRow], source: sourceObject };
+      updatedRow.sourceInvalid = false;
+      updatedRows[activeRow] = updatedRow;
+      setRows(updatedRows);
+    };
+
+    // Assign the setSourceForFilter function to the window object
+    windowRef.setSourceForFilter = setSourceForFilter;
+    
+  }, [rows, activeRow]);
+
+  /**
+    * Handles removing row
+    * @param {object} e : click event
+    * @param {number} index : row index
+  */
+  const handleRemoveRow = (evt: any, index: any) => {
+    const updatedRows = rows.slice(0, index).concat(rows.slice(index + 1));
+    setRows(updatedRows);
+    evt.stopPropagation();
+  };
+
+  /**
+    * handles clearing filter
+    * @param {object} e : click event
+    * @param {number} index : row index
+  */
+  const handleClearFilter = (e: any, index: number) => {
+    const updatedRows = [...rows];
+    updatedRows[index] = {
+      ...defaultValues
+    };
+    setRows(updatedRows);
+  };
+
+  /**
+    * Handles change of every field in the table
+    * @param {number} index : row index
+    * @param {string} field : column name
+    * @param {any} value : changed value
+  */
+  const handleChange = (index: number, field: string, value: any) => {
+    const updatedRows = [...rows];
+    const updatedRow = { ...updatedRows[index] };
+
+    if (field === 'stageFrom') {
+      updatedRow.stageFrom = { ...updatedRow.stageFrom, Name: value };
+
+      const pos = stageFrom.findIndex((obj: any) => obj.Name === updatedRow.stageFrom.Name);
+      updatedRow.toStageOptions = stageFrom.slice(pos);
+      updatedRow.stageTo = { ...updatedRow.stageTo, Name: updatedRow.toStageOptions[0].Name, Ordinal: updatedRow.toStageOptions[0].Ordinal };
+    } else if (field === 'stageTo') {
+      updatedRow.stageTo = { ...updatedRow.stageTo, Name: value };
+    } else {
+      updatedRow[field] = value;
+    }
+
+    updatedRows[index] = updatedRow;
+    setRows(updatedRows);
+  };
+
+  const handleActiveRow = (e: any, index: any) => {
+    setActiveRow(index);
+  };
+
+  const columns: any = useMemo(
```

Comment:
Please explain what the structure of this array of objects is and how it relates to the next constant (`data`) defined below.

There might be a type for it in react documentation that we can use instead of `any[]`, maybe something like `ReactTableColumn[]`.